contrib/endpoints/src/api_manager/config_manager_impl.cc

```diff
@@ -0,0 +1,313 @@
+const int kInceptionFetchRetries = 5;
+// Default rollouts refresh interval in ms
+const int kConfigUpdateCheckInterval = 60000;
+
+const char kRolloutStrategyManaged[] = "managed";
+// Default service management API url
+const char kServiceManagementService[] =
+    "https://servicemanagement.googleapis.com";
+const char kServiceManagementServiceManager[] =
+    "/google.api.servicemanagement.v1.ServiceManager";
+// config_id validation
+const char kValidConfigDateFormat[] = "%Y-%m-%d";
+const char kConfigRevisionDelimeter = 'r';
+// static configs for error handling
+static std::vector<std::pair<std::string, int>> empty_configs;
+
+// Check config_id is valid
+bool isValidConfigId(const std::string& config_id) {
+  if (config_id.length() < 12) {
+    return false;
+  }
+
+  if (config_id[10] != kConfigRevisionDelimeter) {
+    return false;
+  }
+
+  struct tm tm;
+  std::string date_part = config_id.substr(0, 10);
+  if (!strptime(date_part.c_str(), kValidConfigDateFormat, &tm)) {
+    return false;
+  }
+
+  std::string revision_part = config_id.substr(11);
+  if (std::stoi(revision_part) == 0 && revision_part.compare("0") != 0) {
+    return false;
+  }
+
+  return true;
+}
+};
+
+ConfigManagerImpl::ConfigManagerImpl(
+    std::shared_ptr<context::GlobalContext> global_context,
+    std::function<void(const utils::Status&,
+                       std::vector<std::pair<std::string, int>>&)>
+        config_roollout_callback)
+    : global_context_(global_context),
+      config_roollout_callback_(config_roollout_callback),
+      service_management_url_(kServiceManagementService),
+      refresh_interval_ms_(kConfigUpdateCheckInterval) {
+  if (global_context_->server_config()->has_service_management_config()) {
+    // update ServiceManagement service API url
+    if (!global_context_->server_config()
+             ->service_management_config()
+             .url()
+             .empty()) {
+      service_management_url_ =
+          global_context_->server_config()->service_management_config().url();
+    }
+    // update refresh interval in ms
+    if (global_context_->server_config()
+            ->service_management_config()
+            .refresh_interval_ms() > 0) {
+      refresh_interval_ms_ = global_context_->server_config()
+                                 ->service_management_config()
+                                 .refresh_interval_ms();
+    }
+  }
+
+  // Callback to fetch metadata
+  on_fetch_metadata_callback_ = get_fetch_metadata_callback();
+
+  // Calllback to fetch auth token
+  on_fetch_auth_token_callback_ = get_fetch_auth_token_callback();
+}
+
+// check if either service_name or config_id is empty, call fetch_metadata to
+// get it.
+void ConfigManagerImpl::Init() {
+  if (global_context_->service_name().empty() ||
+      global_context_->config_id().empty()) {
+    GlobalFetchGceMetadata(global_context_, on_fetch_metadata_callback_);
+  } else {
+    on_fetch_metadata_callback_(utils::Status::OK);
```

Comment:
this code look strange,  it is NOT easy to read,  it is prefer to call the function directly.   you can do this

void ConfigManagerImpl::on_fetch_metadata_done( utils::Status status) {
}

For 

void ConfigManagerImpl::Init() {
      if   (...) {
           GlobalFetchGceMetadata(,   [this](status) {
                       on_fetch_metadata_done(status); });
    } else {
        on_fetch_metadata_done(OK);
    }
}